File: Java/Leetcode/Greedy/GasStation.java
package com.brianprojects.Leetcode.Greedy;

public class GasStation {
    public int canCompleteCircuit(int[] gas, int[] cost) {

        int totalGas=0, totalCost = 0;
        int[] res = new int[gas.length];
        for(int i =0; i<gas.length;i++){
            res[i] = gas[i]-cost[i];
            totalGas+=gas[i];
            totalCost+=cost[i];
        }
        if(totalGas < totalCost){return -1;}
        int i =0;
        int j = i;
        int gasLeft = res[i];
        while(i<res.length && j<res.length-1){
            //if(j==i){return i;}
            if(res[i] <0 || gasLeft <0){
                gasLeft = res[(i+1)%res.length];
                i+=1;
                j=i;
                //j = (i+1)%res.length;
            }
            else{
                j+=1;
                gasLeft+=res[j];
            }
        }
        return gasLeft>=0?i:-1;
    }


}


/*
for(int i = 0; i<gas.length;i++){
            if(!(cost[i]>gas[i])){
                int left = gas[i]-cost[i];
                int j = (i+1)%gas.length;
                while(left>=0 && j!=i){
                    left+=gas[j];
                    left-=cost[j];
                    j=(j+1)%gas.length;
                }
                if(left >= 0){return i;}
            }
        }
 */

File: Python/Leetcode_Python/Greedy/GasStation.py
# Time Complexity: O(n) || Space Complexity: O(n)
from typing import List


def canCompleteCircuit(gas: List[int], cost: List[int]) -> int:
    if not gas or not cost:
        return -1

    total_gas = sum(gas)

    total_cost = sum(cost)
    if total_cost > total_gas:
        return -1

    curr_i = 0
    curr_sum = gas[0] - cost[0]
    for i in range(1, len(gas)):
        if curr_sum < 0:
            curr_i = i
            curr_sum = 0
        curr_sum += gas[i] - cost[i]

    if curr_sum >= 0:
        return curr_i

    return -1


canCompleteCircuit([1, 2, 3, 4, 5], [3, 4, 5, 1, 2])
